server/src/main/java/org/elasticsearch/common/lucene/Lucene.java

```diff
@@ -869,14 +875,32 @@ public CacheHelper getReaderCacheHelper() {
                 return null; // Modifying liveDocs
             }
         }
+
         DirectoryReaderWithAllLiveDocs(DirectoryReader in) throws IOException {
-            super(in, new FilterDirectoryReader.SubReaderWrapper() {
+            super(in, new SubReaderWrapper() {
                 @Override
                 public LeafReader wrap(LeafReader leaf) {
-                    return new SubReaderWithAllLiveDocs(leaf);
+                    SegmentReader segmentReader = segmentReader(leaf);
+                    Bits hardLiveDocs = segmentReader.getHardLiveDocs();
+                    if (hardLiveDocs == null) {
+                        return new LeafReaderWithLiveDocs(leaf, null, leaf.maxDoc());
+                    }
+                    final int numDocs;
+                    if (hardLiveDocs instanceof FixedBitSet) {
```

Comment:
does it ever kick in? I would assume you get a `FixedBits` instance for live docs instead of a mutable FixedBitSet.

Reply:
Yes, we convert then return a read-only bits. I removed this 💯.